In [1]:
import pandas as pd
import json
import ast
import numpy as np 

In [2]:
# Ajustar la opción para mostrar más caracteres en cada columna
pd.set_option('display.max_colwidth', None)

1 - Lectura de Archivos

In [4]:
# Abrir el archivo JSON y leer línea por línea
with open(r'C:\\Users\\Pablo\\OneDrive\\Henry\\ProyectoFinal\\Google Maps\\metadata-sitios\\1.json') as file:
    data = [json.loads(line) for line in file]

# Convertir la lista de diccionarios en un DataFrame
df1 = pd.DataFrame(data)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\2.json'
# Leer el JSON en un DataFrame
df2 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\3.json'
# Leer el JSON en un DataFrame
df3 = pd.read_json(ruta_json, lines=True)

# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\4.json'
# Leer el JSON en un DataFrame
df4 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\5.json'
# Leer el JSON en un DataFrame
df5 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\6.json'
# Leer el JSON en un DataFrame
df6 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\7.json'
# Leer el JSON en un DataFrame
df7 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\8.json'
# Leer el JSON en un DataFrame
df8 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\9.json'
# Leer el JSON en un DataFrame
df9 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\10.json'
# Leer el JSON en un DataFrame
df10 = pd.read_json(ruta_json, lines=True)


# Ruta al archivo JSON
ruta_json = r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\Google Maps\metadata-sitios\11.json'
# Leer el JSON en un DataFrame
df11 = pd.read_json(ruta_json, lines=True)

2 - Funciones para el analisis

In [3]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address', 
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está en el penúltimo lugar de la dirección y lo almacena en una nueva columna 
    llamada 'city'. Si no hay penúltimo lugar, extrae el último.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene 
               el valor en el penúltimo lugar de 'address', o el último si no hay penúltimo.
    """
    
    # Dividir la columna 'address' en partes usando la coma como delimitador
    df['city'] = df['address'].apply(
        lambda x: x.split(',')[-2].strip() if isinstance(x, str) and len(x.split(',')) > 1 
        else (x.split(',')[-1].strip() if isinstance(x, str) and len(x.split(',')) == 1 else None)
    )
    
    return df


In [4]:
def nulosYVacios(df):
    """
    Este procedimiento recibe un DataFrame y muestra la cantidad de valores nulos, vacíos y la cantidad de 
    valores duplicados en todas las columnas excepto en las 6, 10, 11 y 13 del DataFrame. También muestra 
    el mínimo, máximo y media para columnas de tipo float e integer.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    print("Informe de Nulos, Vacíos y Duplicados:")
    print("-" * 30)

    # Total de registros en el DataFrame
    total_registros = len(df)

    # Imprimir total de registros
    print(f"Total de registros: {total_registros}")
    print("-" * 30)

    # Iterar sobre todas las columnas para mostrar nulos y vacíos
    for columna in df.columns:
        # Contar valores nulos
        nulos = df[columna].isnull().sum()

        # Contar valores vacíos (considerando que solo se cuentan si son cadenas)
        vacios = df[columna].apply(lambda x: isinstance(x, str) and x == '').sum()

        # Calcular porcentajes
        porcentaje_nulos = (nulos / total_registros) * 100 if total_registros > 0 else 0
        porcentaje_vacios = (vacios / total_registros) * 100 if total_registros > 0 else 0

        # Imprimir resultados
        print(f"Columna: {columna}")
        print(f"  Valores Nulos: {nulos}")
        print(f"  Valores Vacíos: {vacios}")
        print(f"  Porcentaje Nulos: {porcentaje_nulos:.2f}%")
        print(f"  Porcentaje Vacíos: {porcentaje_vacios:.2f}%")

        # Contar y mostrar la cantidad de valores duplicados solo si no es una de las columnas con datos anidados
        if columna not in [df.columns[6], df.columns[10], df.columns[11], df.columns[13]]:
            # Filtrar los valores no nulos antes de contar duplicados
            cantidad_duplicados = df[columna].dropna().duplicated(keep=False).sum()
            print(f"  Cantidad de valores duplicados (solo no nulos): {cantidad_duplicados}")

        # Mostrar mínimo, máximo y media si la columna es de tipo float o int
        if df[columna].dtype in ['float64', 'int64']:
            minimo = df[columna].min()
            maximo = df[columna].max()
            media = df[columna].mean()
            print(f"  Mínimo: {minimo}")
            print(f"  Máximo: {maximo}")
            print(f"  Media: {media:.2f}")

        print("-" * 30)


In [5]:
def rankingFrecuenciaCiudad(df, x=10):
    """
    Este procedimiento recibe un DataFrame y un número entero x.
    Muestra en pantalla el ranking de los primeros x elementos con mayor frecuencia en la columna 'city'.

    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    x (int): El número de elementos a mostrar en el ranking.
    """
    print(f"Ranking de los primeros {x} elementos más frecuentes en la columna 'city':")
    print("-" * 60)

    # Verifica si la columna 'city' existe en el DataFrame
    if 'city' in df.columns:
        # Obtener la frecuencia de cada valor en la columna 'city'
        frecuencias = df['city'].value_counts()

        # Obtener los primeros x elementos con mayor frecuencia
        ranking = frecuencias.head(x)

        # Mostrar el ranking
        for index, value in ranking.items():
            print(f"{index}: {value}")
    else:
        print("La columna 'city' no existe en el DataFrame.")

    print("-" * 60)

In [6]:
def mostrarCantidadFilasDuplicadas(df):
    """
    Filtra las filas duplicadas en un DataFrame según columnas específicas,
    y muestra la cantidad de filas duplicadas en la salida de pantalla.
  
    Args:
    df (pd.DataFrame): El DataFrame a analizar.
    """
    # Filtrar duplicados considerando las columnas específicas
    filas_duplicadas = df[df.duplicated(subset=[df.columns[i] for i in [0, 1, 2, 4, 5]], keep=False)]
    
    # Contar la cantidad de filas duplicadas
    cantidad_duplicados = len(filas_duplicadas)
    
    # Mostrar la cantidad de filas duplicadas
    print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

In [7]:
def encontrarDuplicados(df, columnas_excluir=None):
    """
    Encuentra filas duplicadas en un DataFrame basándose en todas las columnas, 
    excepto algunas (las que tienen datos anidados) que pueden ser excluidas.
    
    Args:
        df (pd.DataFrame): El DataFrame donde se buscarán duplicados.
        columnas_excluir (list, optional): Lista de nombres o índices de columnas a excluir del chequeo 
                                           de duplicados. Si es None, se excluyen las columnas 6, 10, 11 y 13.
        
    Returns:
        pd.DataFrame: Un DataFrame que contiene solo las filas duplicadas. 
    """
    # Definir columnas a excluir por defecto si no se proporcionan
    if columnas_excluir is None:
        columnas_excluir = [6, 10, 11, 13]
    
    # Crear una lista de todas las columnas menos las excluidas
    columnas_incluir = df.columns.difference(df.columns[columnas_excluir])
    
    # Verificar duplicados solo en las columnas seleccionadas
    duplicados = df.duplicated(subset=columnas_incluir, keep=False)
    
    # Retornar filas duplicadas
    return df[duplicados]


In [8]:
def eliminarDuplicados(df, columnas_excluir=None):
    """
    Crea una copia del DataFrame excluyendo solo las filas duplicadas, conservando la primera aparición de cada fila duplicada.
    
    Args:
        df (pd.DataFrame): El DataFrame original.
        columnas_excluir (list, optional): Lista de nombres o índices de columnas a excluir del chequeo 
                                           de duplicados. Si es None, se excluyen las columnas 6, 10, 11 y 13.
        
    Returns:
        pd.DataFrame: Una copia del DataFrame sin las filas duplicadas (se conserva la primera aparición).
    """
    # Definir las columnas a excluir por defecto si no se proporcionan
    if columnas_excluir is None:
        columnas_excluir = [6, 10, 11, 13]
    
    # Crear una lista de todas las columnas menos las excluidas
    columnas_incluir = df.columns.difference(df.columns[columnas_excluir])
    
    # Identificar las filas duplicadas, pero conservando la primera aparición de cada duplicado
    duplicados = df.duplicated(subset=columnas_incluir, keep='first')
    
    # Crear una copia del DataFrame excluyendo las filas duplicadas (segundas apariciones en adelante)
    df_sin_duplicados = df[~duplicados]
    
    return df_sin_duplicados


In [10]:
def desanidarColumna(df, col_index):
    """
    Explota la columna especificada por su índice en el DataFrame 
    y mantiene la columna 'gmap_id' en el nuevo DataFrame. 
    También elimina la columna explotada del DataFrame original.
    
    Args:
    df (pd.DataFrame): DataFrame que contiene la columna a desanidar.
    col_index (int): Índice de la columna a desanidar.
    
    Returns:
    tuple: (pd.DataFrame: DataFrame original actualizado, pd.DataFrame: Nuevo DataFrame con 'gmap_id' y la columna explotada)
    """
    # Obtener el nombre de la columna a partir de su índice
    col_name = df.columns[col_index]
    
    # Crear un nuevo DataFrame desanidado con explode
    df_exploded = df.explode(col_name).reset_index(drop=True)
    
    # Seleccionar solo las columnas 'gmap_id' y la columna explotada
    df_result = df_exploded[['gmap_id', col_name]]
    
    # Eliminar la columna especificada del DataFrame original
    df.drop(columns=[col_name], inplace=True)
    
    return df, df_result


In [11]:
def extractCity(df):
    """
    Esta función toma un DataFrame que contiene una columna llamada 'address', 
    la cual tiene direcciones formateadas con comas. La función extrae el valor
    que está en el penúltimo lugar de la dirección y lo almacena en una nueva columna 
    llamada 'city'. Si no hay penúltimo lugar, extrae el último.

    Parámetros:
    df (DataFrame): El DataFrame que contiene la columna 'address' y 'cantidadSplit'.

    Retorna:
    DataFrame: El DataFrame original con una nueva columna 'city' que contiene 
               el valor en el penúltimo lugar de 'address', o el último si no hay penúltimo.
    """
    
    # Dividir la columna 'address' en partes usando la coma como delimitador
    # Extraer el penúltimo elemento o el último elemento si no hay penúltimo
    df['city'] = df['address'].str.split(',').apply(
        lambda x: x[-2].strip() if len(x) > 1 else (x[-1].strip() if len(x) == 1 else None) if x is not None else None
    )
    
    return df



In [18]:
def agregarCantidadSplit(df, columna):
    """
    Agrega una nueva columna al DataFrame que cuenta el número de elementos 
    en una columna específica, dividiendo los valores por comas.

    Parámetros:
    df (DataFrame): El DataFrame al que se le añadirá la nueva columna.
    columna (str): El nombre de la columna que se quiere dividir y contar.

    Retorna:
    DataFrame: El DataFrame original con la nueva columna 'cantidadSplit'.
    """
    # Aplicar una función a cada elemento de la columna especificada
    df['cantidadSplit'] = df[columna].apply(
        lambda x: len(x.split(',')) if isinstance(x, str) else 0
    )
    return df


Frecuencia de recomendaciones por ciudad segun google Maps

Unir todos los Data Frames:

In [12]:
# Lista de DataFrames a unir:
dataframes = [df1, df2, df3, df4, df5, df6, df7, df8, df9, df10, df11]

# Unir todos los DataFrames
df_unido = pd.concat(dataframes, axis=0)

# Guardar el DataFrame unido como archivo Parquet - ARCHIVO INTERMEDIO
df_unido.to_parquet('df_unido.parquet')


In [13]:
# Leer archivo parquet C:\Users\Pablo\OneDrive\Henry\ProyectoFinal
df_unido = pd.read_parquet(r'C:\Users\Pablo\OneDrive\Henry\ProyectoFinal\df_unido.parquet')

In [14]:
df_unido.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3025011 entries, 0 to 275000
Data columns (total 16 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   category          object 
 7   avg_rating        float64
 8   num_of_reviews    int64  
 9   price             object 
 10  hours             object 
 11  MISC              object 
 12  state             object 
 13  relative_results  object 
 14  url               object 
 15  city              object 
dtypes: float64(3), int64(1), object(12)
memory usage: 392.3+ MB


In [23]:
# Verificar el contenido de Addres para extrae la ciudad.
df_unido.iloc[9:20,1:2]
# La columna almacena la ciudad, existen datos nulos.

address
9                                                                                                     None
10                                              Lush Family Dental, 2505 SE Encompass Dr, Waukee, IA 50263
11                                               515 Auto Shades, 5770 Metro E Dr, Pleasant Hill, IA 50327
12                                   Lena’s Beauty Nails Lashes Brows, 657 NE Venture Dr, Waukee, IA 50263
13                Black Tie Ski Rental Delivery of Mammoth, 501 Old Mammoth Rd #9, Mammoth Lakes, CA 93546
14                                                   NTK OUTDOOR, 2315 NW 107th Ave #1B18, Miami, FL 33172
15                                                     Angel-A Massage, 5329 Dewey Dr, Fair Oaks, CA 95628
16                                        Dunn-Edwards Paints, 1850 Douglas Blvd #200, Roseville, CA 95661
17                                                    Elite Massage, 8121 Madison Ave, Fair Oaks, CA 95628
18                                           PODS Sacramento Hub, 5000 Patrol Rd, McClellan Park, CA 95652
19  Happy Spa & Massage (Formerly Hawaii Massage & Spa), 910 Pleasant Grove Blvd #150, Roseville, CA 95678

In [ ]:
# Se separa la columna 'address' en partes separadas por comas, para extraer la ciudad.
agregarCantidadSplit(df_unido,'address')

La columna 'address' contiene las ciudades. Se realiza el siguiente filtrado

In [64]:
# Se cuanta la cantidad de campos que tiene columna cuando la dividimos usando comas:
df_unido['cantidadSplit'].value_counts()

cantidadSplit
4.0     2584966
5.0      215344
3.0       92359
2.0       29372
6.0       17711
7.0        3207
8.0         722
1.0         403
9.0         234
10.0         72
11.0         29
12.0         20
15.0         12
13.0         12
17.0         11
14.0          7
18.0          6
16.0          4
19.0          2
26.0          2
20.0          1
21.0          1
24.0          1
29.0          1
31.0          1
Name: count, dtype: int64

In [69]:
# Ajustar la configuración para mostrar más filas
pd.set_option('display.max_rows', 100)

In [78]:
# Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 8 partes. Se encuentra en el pen ultimo lugar.
print(df_unido[df_unido['cantidadSplit'] == 8.0]['address'])

4275                                  Celebrity Cutz Barbershop - Barber Shop, Barber Hairstyle, Straight Razor, Barber in Alexandria, LA, 3437 Masonic Dr #1584, Alexandria, LA 71301
11303                                                                                    Reagor Air Conditioning, Heating, Plumbing, & Electric, LLC, 300 N Water St, Burnet, TX 78611
15101                                          The Open Door Christian Church, WORSHIP ADDRESS:, MAILING ADDRESS: PO Box 743, New London, MN 56273, 12720 51st St NE, Spicer, MN 56288
15531                                  The Mobile Notary1. Little River, North Myrtle Beach, Calabash North Carolina, Myrtle Beach, Horry County area, 286 Crystal Ln, Longs, SC 29568
15918                                                                       UCHealth - Jennifer Holmes MD, MS, Pavilion 1, Suite 402, 1500, Park Central Dr, Highlands Ranch, CO 80129
                                                                                     

In [77]:
# Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 7 partes. Se encuentra en el pen ultimo lugar.
print(df_unido[df_unido['cantidadSplit'] == 7.0]['address'])

1088                             Gig Harbor Acupuncture & Holistic Health - Kirsty M. Docken, MSOM, EAMP, LAc, 2703 Jahn Ave NW c5, Gig Harbor, WA 98335
1118                                                   Larisa Lazarev, GRI, MRP, PSA | Shorewood Real Estate, 6189 Lehman Dr, Colorado Springs, CO 80918
2611                                                           Sharon H Rhodes, RN, MSN, FNP-BC, 100 Constitution Dr Suite 217, Virginia Beach, VA 23462
3051                                      Margaret Ellen Nichols, MD, Medical Building, 1905 West 32nd Street, Suite 403 Thousand Oaks, Joplin, MO 64804
4626                                                                GBI INK: SIGNS, T-SHIRTS, WRAPS, PRINTING, 1414 S Frazier St # 107, Conroe, TX 77301
                                                                               ...                                                                      
270137                                                     Freedom Financial Netwo

In [76]:
# Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 6 partes. Se encuentra en el pen ultimo lugar.
print(df_unido[df_unido['cantidadSplit'] == 6.0]['address'])

518                             SERVPRO of Benton, Jasper, Newton & Southern Lake Counties, 9582 N Industrial Dr, St John, IN 46373
598                                                                    Ed Khoury, MD, FACOG, 111 Broadway #1405, New York, NY 10006
756       JDM Mafia Motorsports - JDM Parts - Autoparts Exterior , Engine , Suspension, 3590 Redondo Beach Blvd, Torrance, CA 90504
778                                              Matthew F. Wachtler, DPM, FACFAS, 886 Pompton Ave Suite A-1, Cedar Grove, NJ 07009
800                                                     CRST WIC Office, 16618 243rd Avenue, Eagle Butte, SD, Eagle Butte, SD 57625
                                                                    ...                                                            
273382                                                   Stiles Heating, Cooling, and Plumbing, 140 Ben Burton Rd, Bogart, GA 30622
273433                                                             Eugene Ja

In [72]:
# # Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 5 partes. Se encuentra en el pen ultimo lugar.
df_unido[df_unido['cantidadSplit'] == 5.0]['address']


8                                                                     Sweet Rewards Gluten Free Bakery, LLC, 85 NE Dartmoor Dr, Waukee, IA 50263
27                                                       Basinger Family Dentistry: Brian Basinger, DDS, 818 Pierremont Rd, Shreveport, LA 71106
54                                                     Tax Deferred Solutions (TDS Group, INC), 6939 Sunrise Blvd #250, Citrus Heights, CA 95610
82                                                   Regus - California, Irvine - Oracle Tower, 17901 Von Karman Ave Suite 600, Irvine, CA 92614
85                                                                                       Birdi Systems, Inc., 723 E Green St, Pasadena, CA 91101
                                                                           ...                                                                  
274864    Tiedyehobo, 81 - 6592 mamalahoa hwy 2nd location * 92 - 8581 orchid parkway, oceanview, 96737 Kealakekua Is the 1st loca

In [73]:
# Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 3 partes. Se encuentra en el pen ultimo lugar.
df_unido[df_unido['cantidadSplit'] == 3.0]['address']

42                                                                     Knight Hawk, Lyons, KS 67554
44                                                     State Street/7th Street, El Centro, CA 92243
46                                                       Burnette Baptist Church, Detroit, MI 48227
78                                                 Vancouver (99th St Station), Vancouver, WA 98665
160                                     Bixler Lake Disc Golf Course - East, Kendallville, IN 46755
                                                    ...                                            
274789                                 Loving Care In Home Registry Services, Joshua Tree, CA 92252
274879                                                            Thrift Center, Parkland, WA 98444
274918    Wisconsin Historical Marker #425 The Brotherstown Indians of Wisconsin, Chilton, WI 53014
274932                                        Go Fresh Express Wildside, Jackson Township, NJ 08527


In [74]:
# Se verifica en que posicion se encuentra la ciudad cuando se puede dividir en 2 partes. Se encuentra en el pen ultimo lugar.
df_unido[df_unido['cantidadSplit'] == 2.0]['address']

176               Hayes Lake, Washington 98244
411                 Tucannon River, Washington
556                Cabin Creek, Oklahoma 74023
897                    Turkey Creek, Tennessee
961                Kelly Canyon, Arizona 86001
                          ...                 
274138                   Piney River, Virginia
274201             Blue Earth River, Minnesota
274810       1389 Water Well, New Mexico 88030
274811    George Washington Rock, Nevada 89510
274993                 Tyee Campground, Oregon
Name: address, Length: 29372, dtype: object

In [90]:
# Añadir la ciudad a la que pertenecen los comercios en una nueva columna.
extractCity(df_unido)

name  \
0                      Porter Pharmacy   
1                         City Textile   
2                         San Soo Dang   
3                         Nova Fabrics   
4                     Nobel Textile Co   
...                                ...   
274996                        Steak 48   
274997   Jack Mcnerney Chevrolet, Inc.   
274998  Central Ny Spay Neuter Assista   
274999                   Ok Feed Store   
275000              Crestview Crossing   

                                                                    address  \
0                       Porter Pharmacy, 129 N Second St, Cochran, GA 31014   
1                     City Textile, 3001 E Pico Blvd, Los Angeles, CA 90023   
2                    San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005   
3                       Nova Fabrics, 2200 E 11th St, Los Angeles, CA 90021   
4                     Nobel Textile Co, 719 E 9th St, Los Angeles, CA 90021   
...                                                                     ...   
274996                     Steak 48, 260 S Broad St, Philadelphia, PA 19102   
274997           Jack Mcnerney Chevrolet, Inc., 363 NY-281, Tully, NY 13159   
274998  Central Ny Spay Neuter Assista, 17 Salisbury St, Cortland, NY 13045   
274999                Ok Feed Store, 16300 SW 296th St, Homestead, FL 33033   
275000                                Crestview Crossing, Newberg, OR 97132   

                                      gmap_id description   latitude  \
0       0x88f16e41928ff687:0x883dad4fd048e8f8        None  32.388300   
1       0x80c2c98c0e3c16fd:0x29ec8a728764fdf9        None  34.018891   
2       0x80c2c778e3b73d33:0xbdc58662a4a97d49        None  34.058092   
3        0x80c2c89923b27a41:0x32041559418d447        None  34.023669   
4       0x80c2c632f933b073:0xc31785961fe826a6        None  34.036694   
...                                       ...         ...        ...   
274996  0x89c6c74f43a49b55:0x6be6995921c58b12        None  39.947254   
274997  0x89da1787d3fdc4a5:0xce92ea6e3cd4d1cc        None  42.788636   
274998  0x89da6b9ce832726f:0xbd0d3630f1e3b42e        None  42.603831   
274999  0x88d9e872a9e39f3d:0x5de29f50507e5d73        None  25.491641   
275000  0x54956b3944b1654d:0x7797cead9a485c22        None  45.310254   

         longitude                                                   category  \
0       -83.357100                                                 [Pharmacy]   
1      -118.215290                                         [Textile exporter]   
2      -118.292130                                        [Korean restaurant]   
3      -118.232930                                             [Fabric store]   
4      -118.249421                                             [Fabric store]   
...            ...                                                        ...   
274996  -75.164953  [Steak house, Fine dining restaurant, Seafood restaurant]   
274997  -76.122120                                         [Chevrolet dealer]   
274998  -76.165286                                             [Veterinarian]   
274999  -80.456076                                        [Animal feed store]   
275000 -122.935500                                                  [Lodging]   

        avg_rating  num_of_reviews price  \
0              4.9              16  None   
1              4.5               6  None   
2              4.4              18  None   
3              3.3               6  None   
4              4.3               7  None   
...            ...             ...   ...   
274996         4.6             308  None   
274997         4.4             143  None   
274998         4.3              88  None   
274999         4.9              13  None   
275000         5.0               1  None   

                                                                                                                                                               hours  \
0                     [[Friday, 8AM–6PM], [Sa

In [34]:
# Ajustar la opción para mostrar más filas
pd.set_option('display.max_rows', 500)  

Existen 80914 filas sin direccion. Sin embargo tienen la ubicacion (Coordenadas). A traves de las latitudes y longitudes se intento geolocalizar las ciudades a la que pertencian. Sin embargo por una limintacion en el uso de Geocoder no se pudo realizar dicha tarea.

In [101]:
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut

In [105]:
# Crear una instancia del geocodificador
geolocator = Nominatim(user_agent="geoapiExercises")

In [108]:
def obtener_ciudad(lat, lon):
    """
    Esta función toma la latitud y longitud como argumentos y utiliza 
    el servicio de geocodificación para obtener la ciudad correspondiente.

    Args:
        lat (float): La latitud de la ubicación.
        lon (float): La longitud de la ubicación.

    Returns:
        str: El nombre de la ciudad o None si no se encuentra.
    """
    try:
        # Realiza la búsqueda inversa de la ubicación
        location = geolocator.reverse((lat, lon), exactly_one=True)
        
        # Verifica si se encontró una ubicación
        if location:
            # Devuelve el nombre de la ciudad o, si no está disponible, el nombre del pueblo
            return location.raw['address'].get('city', None) or location.raw['address'].get('town', None)
    
    # Maneja el caso en el que hay un timeout
    except GeocoderTimedOut:
        return obtener_ciudad(lat, lon)  # Reintentar si hay un timeout

    # Devuelve None si no se encontró ninguna ciudad
    return None

In [ ]:
# Filtrar las filas donde 'city' es null o None
filtro = df_unido['city'].isnull() | df_unido['city'].isna()

# Aplicar la función solo en las filas filtradas
df_unido.loc[filtro, 'city'] = df_unido.loc[filtro].apply(lambda row: obtener_ciudad(row['latitude'], row['longitude']), axis=1)


In [114]:
# Ranking de Ciudades con mas reseñas
rankingFrecuenciaCiudad(df_unido,10)

Ranking de los primeros 10 elementos más frecuentes en la columna 'city':
------------------------------------------------------------
Houston: 32180
New York: 23218
Chicago: 21607
Los Angeles: 18766
Brooklyn: 18479
San Antonio: 16106
Dallas: 14785
Las Vegas: 14552
Miami: 14345
Philadelphia: 13587
------------------------------------------------------------


Quedaron 80914 filas con valores nulos de ciudades. Cabe mecionar que esta columna se deriva de la columna 'address' por cual estos valores nulos tambien se encuentran contenidos entre los 80511 valores nulos de la columna mencionada.

In [110]:
# Obtencion de Datos para el analisis preliminar de calidad por columna
nulosYVacios(df_unido)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 3025011
------------------------------
Columna: name
  Valores Nulos: 37
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 870876
------------------------------
Columna: address
  Valores Nulos: 80511
  Valores Vacíos: 0
  Porcentaje Nulos: 2.66%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 57210
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 53156
------------------------------
Columna: description
  Valores Nulos: 2770722
  Valores Vacíos: 0
  Porcentaje Nulos: 91.59%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 187002
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nu

In [29]:
# Filtrar las filas donde 'name' es nulo
df_nulos = df_unido[df_unido['name'].isnull()]

# Imprimir el DataFrame filtrado
print(df_nulos)

        name address                                gmap_id description  \
64910   None    None   0x8890d80ebf3e4d95:0xff3ef59fb13f4bb        None   
85097   None    None  0x88db4b02abe5e7a3:0x9a7048d487f53831        None   
108421  None    None  0x862181ce2810a867:0xd591f851846f2d70        None   
152707  None    None  0x89e9969ea6a91ff7:0xd40d6f55c9365737        None   
211758  None    None  0x8669b1501c322927:0xf39e466fa9b182d0        None   
223133  None    None  0x88d1b6021af9c271:0xd74681d91cd1e027        None   
244643  None    None  0x80dc0fe3e09cf6b9:0x387be21ed5309be9        None   
29389   None    None  0x863c3f9f038ff67b:0xddbd4880bd2cd5f0        None   
73516   None    None  0x80e93be99f14306b:0x34a1d32c825c4643        None   
83022   None    None  0x89af4bec1797476d:0xa85251024139a39a        None   
101114  None    None  0x54c39ca6043a9b67:0x5304e8b9a74f3421        None   
102189  None    None  0x89c213fa256d41fd:0x52c21e9b8ff4301f        None   
129435  None    None  0x8

Los valores cuyo nombre son nulos, solo tienen las coordenas, algunos de ellos son espacios publicos (Como ciudades o playas) y no hacen referencia al horario en que abren. Aun asi algunos usuarios han realizado una calificacion y comentarios del lugar.

3 - Analisis Preeliminar de datos, primeros hallazgos:
Este informe proporciona un análisis exhaustivo de la calidad de los datos en un DataFrame de 3,025,011 registros. Aquí hay un resumen de los hallazgos más relevantes:

1. **Valores Nulos y Vacíos**:
   - Varias columnas, como `description`, `price`, y `hours`, presentan un alto número de valores nulos. En particular, la columna `description` tiene un 91.59% de valores nulos, lo que indica que la mayoría de las entradas no tienen descripción. En concecuencia estos datos habra que analizar su utilidad en el analisis.
   - La columna `price` también tiene un alto porcentaje de valores nulos (90.90%).
   - La columna `city` contiene 80914 valores nulos.
   - `name` tiene un bajo numero de valores nulos (37).

2. **Duplicados**:
   - La columna `gmap_id` tiene 53,156 valores duplicados, lo que sugiere que hay múltiples registros asociados con la misma identificación de Google Maps. Este tipo de duplicados podría ser un área de preocupación si se necesita unicidad en la identificación de lugares.

3. **Columnas Numéricas**:
   - Las columnas `latitude` y `longitude` tienen un rango amplio de valores, con latitudes que van desde -40.93 hasta 87.86 y longitudes desde -178.81 hasta 180.0. Esto puede ser esperado, pero es importante validar que todos los registros caen dentro de las coordenadas geográficas válidas.
   - La columna `avg_rating` tiene una media de 4.30, lo que sugiere una percepción mayormente positiva en la evaluación de los lugares.

4. **Valores Vacíos**:
   - Hay una cantidad mínima de valores vacíos en las columnas, lo que indica que la mayoría de las entradas están completas en este aspecto. Sin embargo, la columna `city` tiene 687 valores vacíos, lo que podría ser relevante si se realizan análisis geográficos.

5. **Observaciones Generales**:
   - Alta proporción de valores nulos en columnas clave como `description`, `price`, y `MISC`.

6. **Columnas con datos anidados**:
   - `Misc`, `category`, `hours` y `relative_results`



In [34]:
df_nulos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2770722 entries, 0 to 275000
Data columns (total 17 columns):
 #   Column            Dtype  
---  ------            -----  
 0   name              object 
 1   address           object 
 2   gmap_id           object 
 3   description       object 
 4   latitude          float64
 5   longitude         float64
 6   category          object 
 7   avg_rating        float64
 8   num_of_reviews    int64  
 9   price             object 
 10  hours             object 
 11  MISC              object 
 12  state             object 
 13  relative_results  object 
 14  url               object 
 15  city              object 
 16  cantidadSplit     int64  
dtypes: float64(3), int64(2), object(12)
memory usage: 380.5+ MB


In [30]:
# Se procede a eliminar las filas que se encuentran repetidas, considerando el mismo creterio mencionado previamente.
dfLimpio = eliminarDuplicados(df_unido)

# Se procede a eliminar la columna cantidadSplit
dfLimpio = dfLimpio.drop(columns=['cantidadSplit'])


In [124]:
# Guardar sin duplicados
dfLimpio.to_parquet('dfLimpio.parquet', index=False)

In [31]:
nulosYVacios(dfLimpio)

Informe de Nulos, Vacíos y Duplicados:
------------------------------
Total de registros: 2998429
------------------------------
Columna: name
  Valores Nulos: 37
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 823946
------------------------------
Columna: address
  Valores Nulos: 79520
  Valores Vacíos: 0
  Porcentaje Nulos: 2.65%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 6108
------------------------------
Columna: gmap_id
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 0.00%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 2
------------------------------
Columna: description
  Valores Nulos: 2745413
  Valores Vacíos: 0
  Porcentaje Nulos: 91.56%
  Porcentaje Vacíos: 0.00%
  Cantidad de valores duplicados (solo no nulos): 185439
------------------------------
Columna: latitude
  Valores Nulos: 0
  Valores Vacíos: 0
  Porcentaje Nulos: 

Comparando con el informe anterior, las principales diferencias son:

1. **Total de Registros**: El número total de registros ha disminuido de **3,025,011** a **2,998,428**.

2. **Valores Nulos en `address`**: La cantidad de valores nulos en `address` ha disminuido ligeramente, de **80,511** a **79,520**, lo que representa un cambio mínimo en la calidad de esta columna.

3. **Valores Nulos en `description`**: El porcentaje de valores nulos en `description` se mantiene alto, con **2,745,412** valores nulos, lo que equivale al **91.56%** de los registros, sin cambios significativos.

4. **Valores Nulos en `price`**: El número de valores nulos en `price` ha disminuido, de **2,749,808** a **2,724,510**, lo que indica que todavía hay un alto porcentaje de datos faltantes.

5. **Duplicados en `address`**: La cantidad de duplicados en `address` ha disminuido, de **57,210** a **6,180**.

6. **Duplicados en `gmap_id`**: La columna `gmap_id` ahora muestra **0** duplicados, indicando que se ha mantenido la unicidad en esta columna, lo cual es un punto positivo en comparación con la versión anterior.

7. **Valores Nulos en `state`**: El número de valores nulos en `state` ha disminuido, de **746,455** a **738,858**, lo que indica una ligera mejora en la completitud de esta columna.

En resumen, el nuevo informe muestra una ligera mejora en algunas columnas, pero persisten problemas significativos de nulos y duplicados en columnas clave, lo que requiere atención adicional.
Otros como `desciption` o `price`contienen un elevado porcentaje de valores nulos (91% y 90%) seguido por `hours` (26%) y `MISC`

In [41]:
# Filtrar las filas donde alguna de las columnas 'description', 'price' o 'hours' es nula
df_nulos = df_unido[df_unido[['description', 'price', 'hours']].isnull().any(axis=1)]

# Imprimir el DataFrame filtrado
print(df_nulos)


                                  name  \
0                      Porter Pharmacy   
1                         City Textile   
2                         San Soo Dang   
3                         Nova Fabrics   
4                     Nobel Textile Co   
...                                ...   
274996                        Steak 48   
274997   Jack Mcnerney Chevrolet, Inc.   
274998  Central Ny Spay Neuter Assista   
274999                   Ok Feed Store   
275000              Crestview Crossing   

                                                                    address  \
0                       Porter Pharmacy, 129 N Second St, Cochran, GA 31014   
1                     City Textile, 3001 E Pico Blvd, Los Angeles, CA 90023   
2                    San Soo Dang, 761 S Vermont Ave, Los Angeles, CA 90005   
3                       Nova Fabrics, 2200 E 11th St, Los Angeles, CA 90021   
4                     Nobel Textile Co, 719 E 9th St, Los Angeles, CA 90021   
...                  

Desanidar Columnas que contienen lista. Para ello no se trabajara con la totalidad del DF sino con las ciudades mas importantes:

In [130]:
# Lista de las 10 ciudades con más reseñas
ciudades_top_10 = [
    'Houston',
    'New York',
    'Chicago',
    'Los Angeles',
    'Brooklyn',
    'San Antonio',
    'Dallas',
    'Las Vegas',
    'Miami',
    'Philadelphia'
]

# Filtrar el DataFrame para seleccionar solo las filas que corresponden a estas ciudades
dfPrueba = dfLimpio[dfLimpio['city'].isin(ciudades_top_10)]


In [131]:
# Se desanida la columna 'category'
dfPrueba, dfCategory = desanidarColumna(dfPrueba, 6)

C:\Users\Pablo\AppData\Local\Temp\ipykernel_1624\2815466058.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=[col_name], inplace=True)


In [132]:
# Guardar en un archivo .parquet
dfCategory.to_parquet('dfCategory.parquet', index=False)

In [133]:
dfPrueba.info()


<class 'pandas.core.frame.DataFrame'>
Index: 185981 entries, 1 to 274996
Data columns (total 15 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   name              185981 non-null  object 
 1   address           185981 non-null  object 
 2   gmap_id           185981 non-null  object 
 3   description       22928 non-null   object 
 4   latitude          185981 non-null  float64
 5   longitude         185981 non-null  float64
 6   avg_rating        185981 non-null  float64
 7   num_of_reviews    185981 non-null  int64  
 8   price             20114 non-null   object 
 9   hours             144207 non-null  object 
 10  MISC              147249 non-null  object 
 11  state             146100 non-null  object 
 12  relative_results  168662 non-null  object 
 13  url               185981 non-null  object 
 14  city              185981 non-null  object 
dtypes: float64(3), int64(1), object(11)
memory usage: 22.7+ MB


In [134]:
# Se desanida la columna 'relative_results'
dfPrueba, dfRelativeResults = desanidarColumna(dfPrueba, 12)

C:\Users\Pablo\AppData\Local\Temp\ipykernel_1624\2815466058.py:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=[col_name], inplace=True)


In [135]:
# Guardar en un archivo .parquet
dfRelativeResults.to_parquet('dfRelativeResults.parquet', index=False)

In [136]:
dfPrueba.info()

<class 'pandas.core.frame.DataFrame'>
Index: 185981 entries, 1 to 274996
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   name            185981 non-null  object 
 1   address         185981 non-null  object 
 2   gmap_id         185981 non-null  object 
 3   description     22928 non-null   object 
 4   latitude        185981 non-null  float64
 5   longitude       185981 non-null  float64
 6   avg_rating      185981 non-null  float64
 7   num_of_reviews  185981 non-null  int64  
 8   price           20114 non-null   object 
 9   hours           144207 non-null  object 
 10  MISC            147249 non-null  object 
 11  state           146100 non-null  object 
 12  url             185981 non-null  object 
 13  city            185981 non-null  object 
dtypes: float64(3), int64(1), object(10)
memory usage: 21.3+ MB


In [137]:
# Guardar con las 10 ciudades principales como parquet archivo limpio sin Normalizar
dfPrueba.to_parquet('dfPrueba.parquet', index=False)

In [138]:
# Copiar el DataFrame con las columnas 'gmap_id' y 'hours'
df_copy = dfPrueba[['gmap_id', 'hours']].copy()

# Reemplazar None con una lista vacía
df_copy['hours'] = df_copy['hours'].apply(lambda x: [] if x is None else x)

# Crear una lista de días de la semana
days_of_week = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

# Crear un DataFrame vacío para los horarios normalizados
hours_normalized = pd.DataFrame(index=df_copy.index)

# Iterar sobre los días de la semana y llenar el DataFrame normalizado
for day in days_of_week:
    hours_normalized[day] = df_copy['hours'].apply(
        lambda x: next((h[1] for h in x if h[0] == day), None)
    )

# Concatenar el DataFrame original con las columnas normalizadas
df_normalized = pd.concat([df_copy[['gmap_id']], hours_normalized], axis=1)

# Mostrar el nuevo DataFrame normalizado
print(df_normalized.head())

                                  gmap_id   Monday     Tuesday   Wednesday  \
1   0x80c2c98c0e3c16fd:0x29ec8a728764fdf9     None        None        None   
2   0x80c2c778e3b73d33:0xbdc58662a4a97d49   Closed  6:30AM–6PM  6:30AM–6PM   
3    0x80c2c89923b27a41:0x32041559418d447  9AM–5PM     9AM–5PM     9AM–5PM   
4   0x80c2c632f933b073:0xc31785961fe826a6  9AM–5PM     9AM–5PM     9AM–5PM   
14   0x88d9beb4fe0532c1:0xef0555c169299d6  9AM–6PM     9AM–6PM     9AM–6PM   

      Thursday      Friday    Saturday    Sunday  
1         None        None        None      None  
2   6:30AM–6PM  6:30AM–6PM  6:30AM–6PM  7AM–12PM  
3      9AM–5PM     9AM–5PM      Closed    Closed  
4      9AM–5PM     9AM–5PM      Closed    Closed  
14     9AM–6PM     9AM–6PM      Closed    Closed  


In [139]:
# Concatenar dfPrueba y df_normalized por la columna gmap_id
df_final = pd.merge(dfPrueba, df_normalized, on='gmap_id', how='left')

# Eliminar la columna 'hours'
df_final = df_final.drop('hours', axis=1)

In [144]:
# Normalizar la columna 'MISC'
misc_normalized = pd.json_normalize(df_final['MISC'])

# Concatenar el DataFrame original (sin la columna MISC) con el nuevo DataFrame normalizado
df_final = pd.concat([df_final.drop(columns=['MISC']), misc_normalized], axis=1)


In [145]:
# Data Frame para EDA
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185981 entries, 0 to 185980
Data columns (total 37 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   name               185981 non-null  object 
 1   address            185981 non-null  object 
 2   gmap_id            185981 non-null  object 
 3   description        22928 non-null   object 
 4   latitude           185981 non-null  float64
 5   longitude          185981 non-null  float64
 6   avg_rating         185981 non-null  float64
 7   num_of_reviews     185981 non-null  int64  
 8   price              20114 non-null   object 
 9   state              146100 non-null  object 
 10  url                185981 non-null  object 
 11  city               185981 non-null  object 
 12  Monday             144207 non-null  object 
 13  Tuesday            144207 non-null  object 
 14  Wednesday          144207 non-null  object 
 15  Thursday           144207 non-null  object 
 16  Fr

In [150]:
# Guardar con las 10 ciudades principales como parquet archivo limpio sin Normalizar
df_final.to_parquet('df_final.parquet', index=False)

### DataFrame Resumen de Cambios

| Index | Column             | Dtype    | Comentarios                                                      |
|-------|--------------------|----------|------------------------------------------------------------------|
| 0     | name               | object   |                                                                  |
| 1     | address            | object   |                                                                  |
| 2     | gmap_id            | object   | Valor llave para las tablas expandidas de 'category' y 'relative_results'. |
| 3     | description        | object   |                                                                  |
| 4     | latitude           | float64  |                                                                  |
| 5     | longitude          | float64  |                                                                  |
| 6     | category           | object   | Expandida, luego eliminada.                                     |
| 7     | avg_rating         | float64  |                                                                  |
| 8     | num_of_reviews     | int64    |                                                                  |
| 9     | price              | object   |                                                                  |
| 10    | hours              | object   | Normalizada, una columna por cada día.                          |
| 11    | MISC               | object   | Normalizada.                                                   |
| 12    | state              | object   |                                                                  |
| 13    | relative_results    | object   | Expandida, luego eliminada.                                     |
| 14    | url                | object   |                                                                  |
| 15    | city               | object   | Generada a partir de 'address'.                                 |

